coap/api/transport.go

```diff
@@ -66,75 +68,84 @@ func MakeCoAPHandler(svc coap.Service, l *slog.Logger) mux.HandlerFunc {
+func sendResp(w mux.ResponseWriter, resp *pool.Message) {
+	if err := w.Conn().WriteMessage(resp); err != nil {
 		logger.Warn(fmt.Sprintf("Can't set response: %s", err))
 	}
 }
 
 func handler(w mux.ResponseWriter, m *mux.Message) {
-	resp := message.Message{
-		Code:    codes.Content,
-		Token:   m.Token,
-		Context: m.Context,
-		Options: make(message.Options, 0, 16),
+	resp := pool.NewMessage(w.Conn().Context())
+	resp.SetToken(m.Token())
+	for _, opt := range m.Options() {
+		resp.AddOptionBytes(opt.ID, opt.Value)
 	}
-	defer sendResp(w, &resp)
+	defer sendResp(w, resp)
+
 	msg, err := decodeMessage(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error decoding message: %s", err))
-		resp.Code = codes.BadRequest
+		resp.SetCode(codes.BadRequest)
 		return
 	}
 	key, err := parseKey(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error parsing auth: %s", err))
-		resp.Code = codes.Unauthorized
+		resp.SetCode(codes.Unauthorized)
 		return
 	}
-	switch m.Code {
+
+	switch m.Code() {
 	case codes.GET:
-		err = handleGet(m.Context, m, w.Client(), msg, key)
+		resp.SetCode(codes.Content)
+		err = handleGet(m, w, msg, key)
 	case codes.POST:
-		resp.Code = codes.Created
-		err = service.Publish(m.Context, key, msg)
+		resp.SetCode(codes.Created)
+		err = service.Publish(m.Context(), key, msg)
 	default:
-		err = svcerr.ErrNotFound
+		err = errMethodNotAllowed
 	}
+
 	if err != nil {
 		switch {
 		case err == errBadOptions:
-			resp.Code = codes.BadOption
-		case err == svcerr.ErrNotFound:
-			resp.Code = codes.NotFound
-		case errors.Contains(err, svcerr.ErrAuthorization),
-			errors.Contains(err, svcerr.ErrAuthentication):
-			resp.Code = codes.Unauthorized
+			resp.SetCode(codes.BadOption)
+		case err == errMethodNotAllowed:
+			resp.SetCode(codes.MethodNotAllowed)
+		case errors.Contains(err, svcerr.ErrAuthorization):
+			resp.SetCode(codes.Forbidden)
+		case errors.Contains(err, svcerr.ErrAuthentication):
+			resp.SetCode(codes.Unauthorized)
 		default:
-			resp.Code = codes.InternalServerError
+			resp.SetCode(codes.InternalServerError)
 		}
 	}
 }
 
-func handleGet(ctx context.Context, m *mux.Message, c mux.Client, msg *messaging.Message, key string) error {
+func handleGet(m *mux.Message, w mux.ResponseWriter, msg *messaging.Message, key string) error {
 	var obs uint32
-	obs, err := m.Options.Observe()
+	obs, err := m.Options().Observe()
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error reading observe option: %s", err))
 		return errBadOptions
 	}
 	if obs == startObserve {
-		c := coap.NewClient(c, m.Token, logger)
-		return service.Subscribe(ctx, key, msg.GetChannel(), msg.GetSubtopic(), c)
-	}
-	return service.Unsubscribe(ctx, key, msg.GetChannel(), msg.GetSubtopic(), m.Token.String())
+		c := coap.NewClient(w.Conn(), m.Token(), logger)
+		w.Conn().AddOnClose(func() {
+			if err := service.Unsubscribe(context.Background(), key, msg.GetChannel(), msg.GetSubtopic(), c.Token()); err != nil {
```

Comment:
Something like this log will be usefull
```suggestion
			logger.Warn(fmt.Sprintf("Kicking out ideal client token %s subscribed for channel %s and subtopic %s", c.Token(), msg.GetChannel(), msg.GetSubtopic()))
```

Reply:
When the connection is closed should the client be automatically closed or the log message if after unsubscribing?

Reply:
Do you mean "idle", not "ideal"? Also "subscribed to channel", not "for channel". Also, DRY,  check how we handle "with(out) error" constructs in logging middleware of our services.